### PR TITLE
Context Parallel Support

### PR DESCRIPTION
**Pull request by farhadrgh**

### CODE REVIEW

File: nvsubquadratic/modules/distributed.py [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+        _use_cp: Whether context parallel is enabled
+        cp_world_size: Context parallel world size
+
+    Returns:
+        Sliced weight tensor for current CP rank
+    """
+    if not _use_cp or cp_world_size <= 1:
+        return weight
+
+    cp_rank = get_context_parallel_rank()
```
Comment (jstjohn) [inline nvsubquadratic/modules/distributed.py]: would recommend passing the cp group around rather than this. If you have the group you can call `torch.distributed.get_rank(group)` for example, or you could do that upstream and pass the rank in. Since you already pass the cp_world_size in it would prob make sense to also pass the cp rank in and not bother calling out to these global functions here.

File: nvsubquadratic/modules/distributed.py [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+        _use_cp: Whether context parallel is enabled
+        cp_world_size: Context parallel world size
+
+    Returns:
+        Sliced bias tensor for current CP rank
+    """
+    if not _use_cp or cp_world_size <= 1:
+        return bias
+
+    cp_rank = get_context_parallel_rank()
```
Comment (jstjohn) [inline nvsubquadratic/modules/distributed.py]: Same comment here on the cp rank being another arg to this function.

Comment (jstjohn) [reply nvsubquadratic/modules/distributed.py]: The rationale is that if you keep everything top level, it will be a lot easier to refactor later. Like ideally all of the group gathering stuff happens in your top level module (or outside it even) and then you can easily refactor later. That way you can use megatron or device mesh, or some new megatron after they change their API to be more like device mesh. It seems like they are very actively working on refactoring things to move away from these kinds of global functions so it would be nice to get ahead of that and detangle things.

File: nvsubquadratic/modules/distributed.py [hunk 3]
```diff
@@ -0,0 +1,792 @@
...
+    bias_per_rank = bias.shape[0] // cp_world_size
+    start_idx = cp_rank * bias_per_rank
+    end_idx = start_idx + bias_per_rank
+    return bias[start_idx:end_idx]
+
+
+def prepare_depthwise_weights_for_cp(
+    weight: torch.Tensor, actual_in_channels: int, group_dim: int, _use_cp: bool, cp_world_size: int
```
Comment (jstjohn) [inline nvsubquadratic/modules/distributed.py]: Same comment here about passing the cp rank in so you can pass it into the child functions.

File: nvsubquadratic/modules/distributed.py [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+
+
+def prepare_standard_weights_for_cp(
+    weight: torch.Tensor,
+    out_channels: int,
+    group_dim: int,
+    groups: int,
+    kernel_size: Tuple[int, ...],
+    _use_cp: bool,
+    cp_world_size: int,
```
Comment (jstjohn) [inline nvsubquadratic/modules/distributed.py]: cp rank

File: nvsubquadratic/modules/distributed.py [hunk 5]
```diff
@@ -0,0 +1,792 @@
...
+
+        # Initialize using standard Conv1d initialization
+        bounds = math.sqrt(1.0 / (self.in_channels * math.prod(self.kernel_size)))
+        conv_init_method = partial(torch.nn.init.uniform_, a=-bounds, b=bounds)
+        initialize_affine_weight_gpu(self.weight, conv_init_method, partition_dim=0)
+
+        # Initialize bias if enabled
+        if self.bias_enabled:
+            self.bias = nn.Parameter(torch.empty(self.width_per_tp_group, device=self.device, dtype=self.dtype))
+            setattr(self.bias, "tensor_model_parallel", True)
```
Comment (jstjohn) [inline nvsubquadratic/modules/distributed.py]: This is used by megatron I believe during the DDP step. The key thing here is that if you are using TP then some of your weights are expected to be different across the cluster. For example if you had TP=2 then this bias would be cut in half, and you would want to avoid doing the standard mean_reduce(weight, cp+dp group) thing that DDP does. 

I would be careful about this and add in tests that cover the case of TP=2, CP=2, etc and verify that gradients through the model match up. See for example: https://github.com/NVIDIA/bionemo-framework/pull/1219 for some testing functions like this that take advantage of the fact that if you set adam b1,b2 to 0,0 then the optimizer state will contain your gradients, so do one step of training like that and you can see what the gradients look like and compare them with/without model parallelism.

Comment (jstjohn) [reply nvsubquadratic/modules/distributed.py]: Worst case you could write an extension of DDP that handles these magic flags set on TP split parameters and skips the reduction of these parameters across the TP group. You would still want to reduce this TP rank across DDP, but we should check with the megatron team on the best way to order that reduction.

Comment (jstjohn) [reply nvsubquadratic/modules/distributed.py]: megatron FSDP may already do some of this. Check with Cory.

File: tests/torchrun_sequence_mixer_cp_test.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+                assert len(grads_without_cp) == len(grads_with_cp)
+                logging.info(f"Comparing {len(grads_without_cp)} gradient tensors")
+            except AssertionError as e:
+                logging.error(f"Gradient count mismatch: {e}")
+                raise
+
+            gradient_mismatch = False
+            for (n_without_cp, g_without_cp), (n_with_cp, g_with_cp) in zip(grads_without_cp, grads_with_cp):
+                try:
+                    torch.testing.assert_close(g_without_cp, g_with_cp)
```
Comment (jstjohn) [inline tests/torchrun_sequence_mixer_cp_test.py:336]: Also check out the test here: https://github.com/NVIDIA/bionemo-framework/pull/1219 (a lot of the cp resharding stuff looks not necessary since you have easier access to grads?). The other kind of metric you can use is `linalg.norm(g1 - g2) / linalg.norm(g1)`, and then the ttrace paper https://arxiv.org/abs/2506.09280 has some guidance in theorem 5.3 I think around defining expected upper bounds of that metric as a function for layer number. If you are doing it isolated then maybe the layer 0 or N (whichever is more stringent) would be the expected bound.

File: tests/torchrun_sequence_mixer_cp_test.py [hunk 7]
```diff
@@ -0,0 +1,382 @@
+"""Context Parallelism Test for QKVSequenceMixer.
```
Comment (jstjohn) [inline tests/torchrun_sequence_mixer_cp_test.py:3]: Definitely also add tests for `TP` coverage. I'm seeing/remembering that we have to set magic variables on parameters that get `TP` split, and my guess is that the default `DDP` configuration will average over the TP group and mess up your grads (which should be specific/different in each TP rank since the weights are different). You may even use that as a negative control for your test, make sure that your test fails when you use regular DDP and TP, otherwise it's not sensitive enough to pick up bugs at that scale.

**Comment by farhadrgh** [comment]
Dropping TP support in [71e44c2](https://github.com/NVIDIA-Digital-Bio/nvSubquadratic-private/pull/6/commits/71e44c2da2c8d9e9e43db1e859277cd799130fd3) for future reference.

File: nvsubquadratic/modules/distributed.py (hunk 1 above)
Comment (farhadrgh) [reply nvsubquadratic/modules/distributed.py]: done

File: nvsubquadratic/modules/distributed.py (hunk 5 above)
Comment (farhadrgh) [reply nvsubquadratic/modules/distributed.py]: I dropped TP support and will make room for FSDP

File: tests/torchrun_sequence_mixer_cp_test.py (hunk 7 above)
Comment (farhadrgh) [reply tests/torchrun_sequence_mixer_cp_test.py:3]: yeah so I decided to drop TP and only support CP + FSDP (future)

File: tests/torchrun_sequence_mixer_cp_test.py (hunk 6 above)
Comment (farhadrgh) [reply tests/torchrun_sequence_mixer_cp_test.py:336]: Added relative error check to the tests in [fc89390](https://github.com/NVIDIA-Digital-Bio/nvSubquadratic-private/pull/6/commits/fc893905ff4b7f6ffb692e4836e7a9acf5065887)

File: nvsubquadratic/modules/self_attention.py [hunk 8]
```diff
@@ -8,6 +8,7 @@
 from einops import rearrange
 from torch.nn.attention import SDPBackend, sdpa_kernel
 
+from nvsubquadratic.parallel.utils import zigzag_gather_from_group_ranks, zigzag_split_across_group_ranks
```
Comment (jstjohn) [inline nvsubquadratic/modules/self_attention.py:11]: Is there any way you can use the standard pytorch attention block here? Maybe a task for a future PR? That also may make it so that you aren't required to stick to the zigzag style, which the base pytorch implementation may not use. This is what torchtitan does afaik: https://docs.pytorch.org/tutorials/unstable/context_parallel.html

File: nvsubquadratic/parallel/utils.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    """
+    num_gpus = torch.cuda.device_count()
+    required_world_size = tensor_model_parallel_size * pipeline_model_parallel_size * context_parallel_size
+    assert num_gpus == required_world_size, (
+        f"World size {num_gpus} != TP={tensor_model_parallel_size} x PP={pipeline_model_parallel_size} x CP={context_parallel_size}"
+    )
+
+    # Set up environment variables
+    os.environ["TORCH_NCCL_BLOCKING_WAIT"] = "0"
+    os.environ["TORCH_NCCL_ASYNC_ERROR_HANDLING"] = "1"
```
Comment (jstjohn) [inline nvsubquadratic/parallel/utils.py]: What if you only set these if they are unset? something like `if "TORCH_NCCL_BLOCKING_WAIT" not in os.environ: ...`

File: nvsubquadratic/modules/self_attention.py (hunk 8 above)
Comment (jstjohn) [reply nvsubquadratic/modules/self_attention.py:11]: Also this implementation is going to suffer from some pretty significant memory issues at long context due to the attention layers not having internal ring attention. Is the plan that the user wouldn't actually use this layer in a real setting? If so a comment about how this is for illustration only and that a user should use a CP enabled self attention block.

**Review by jstjohn — DISMISSED** [review]

File: nvsubquadratic/modules/self_attention.py (hunk 8 above)
Comment (farhadrgh) [reply nvsubquadratic/modules/self_attention.py:11]: yes, sounds good, I created an issue and added some comments in the code to track this. I think we can take a look after getting a full training loop working with Megatron parallelism.
